README.md

```diff
@@ -74,6 +74,18 @@ Current guess at first steps (very open to debate):
 3. Build next-gen fontmake, converting parts to Rust and enabling parallelism
    * Guessing we would either do a good interchange format ([flatfont](https://github.com/googlefonts/flatfont)?) + ninja OR a Rust based executor using something like Rayon
 
+## Definitions
+
+### zerocopy
+
+As a shorthand, we use the term `zerocopy` to describe some desired properties of the font parsing API:
+
+* Reads of objects `n` bytes in size should consume no more than `n` bytes of font data
```

Comment:
This is too strict, isn't it? `O(n)` is better to me.

Reply:
To me the crucial zerocopy assertion is that given an unstructured view - concretely a byte slice, such as from an mmap'd file - I can get a structured view (literally a struct!) of some part of it w/o causing any copying. Is that a poor way to reason about it?

Reply:
The issue for me is that 'copy' is poorly defined. Is it a copy to do a BE->LE conversion on a u32? If we're creating native structs at all, this involves *some* copying, because we cannot just do some kind of reinterpreting cast of the bytes to a type.

Reply:
> literally casting &[u8] to &SomeType


If SomeType has field with types that speak big endian then no copy occurs. You "pay" to flip bytes for fields you access but not for any of the others. To give a concrete example, if you used the zerocopy crate you'd have things like this:


```
#[repr(C)]
#[derive(FromBytes, AsBytes, Unaligned)]
#[derive(Debug, Clone)]
struct TableHeader {
    raw_tag: U32<BigEndian>,
    checksum: U32<BigEndian>,
    offset: U32<BigEndian>,
    length: U32<BigEndian>,
}
```

Reply:
Hmm, I'm confused by that example.

looking at the source for `zerocopy`'s byteorder types, it sure seems like each of those `U32` is backed by a `[u8; 4]`, which was copied from somewhere? To be 'zerocopy', per my understanding, this would have to look something like:

```rust
struct<'a>TableHeader(&'a [u8]);

// assume that slice has valid length by construction
impl TableHeader<'_> {
    fn tag(&self) -> [u8; 4] {
        self[..4].try_into().unwrap()
    }
    
    fn checksum(&self) -> u32 {
        u32::from_bytes_be(self.0[4..8].try_into().unwrap())
    }

    fn offset(&self) -> u32 {
        u32::from_bytes_be(self.0[8..12].try_into().unwrap())
    }

    fn length(&self) -> u32 {
        u32::from_bytes_be(self.0[12..16].try_into().unwrap())
    }
}
```

Reply:
> as long as the type was behind a reference

Yeah, this is the general idea and it's also how HarfBuzz works. We take pointers to aggregates rather than copying them.

Reply:
> This is too strict, isn't it? `O(n)` is better to me.

I meant to type `O(n)` but think I wanted it to be `n` and succumbed to temptation :)

Reply:
> each of those U32 is backed by a [u8; 4], which was copied from somewhere

The key is it wasn't copied, which I think is what your edit is saying

Reply:
So those two implementations of `TableHeader` are roughly equivalent with the exception that the pinot style version does a bounds check per field access. 

The advantages to the “zerocopy” version are threefold:
1. We can perform a bounds check on the entire struct when doing the pointer cast. 
2. We get mutability for free. 
3. It’s pretty much a direct translation of the spec. 

In both cases, we’re doing the byte swap per field access. 

The one advantage I do see with the second method is API uniformity— accessing everything though methods makes usage feel more consistent. I don’t think this outweighs the benefits of the other approach though.

Reply:
> Example: if you ask the font face how many glyphs it has, that two byte number lives in maxp table. Getting there takes some work...

Thanks for this example. I’m in sync with you now. My intent here is to impose constraints on both accessing data at leaves and performing edge traversals (following offsets). I think if we minimize these two, we’ll also be able to offer a minimum cost path from the font root to any target data?

My expectation is that any consumer that wants better guarantees than we can deliver within these constraints will cache their own data either with the font instance (HarfBuzz) or in some side structure (swash).

Reply:
I like the enumeration of advantages, we should capture that :)

Reply:
In the second case we do not need to bounds check

> So those two implementations of `TableHeader` are roughly equivalent with the exception that the pinot style version does a bounds check per field access.

> The advantages to the “zerocopy” version are threefold:
> 
> 1. We can perform a bounds check on the entire struct when doing the pointer cast.
> 2. We get mutability for free.
> 3. It’s pretty much a direct translation of the spec.
>
> In both cases, we’re doing the byte swap per field access.
> 
> The one advantage I do see with the second method is API uniformity— accessing everything though methods makes usage feel more consistent. I don’t think this outweighs the benefits of the other approach though.

I've been mulling this most of the afternoon.

In the second case, we can also get away with a single bounds check; when we construct the `TableHeader` we ensure that the slice is the correct length, and then we can use `get_unchecked` to grab the specific fields.

I do like the "direct translation of the spec" angle. The problem for me, though, is that this seems to only hold when our types contain only scalars? Once we start dealing with collections then we need some other abstraction. Mixing and matching the two approaches (generating fields for these items, but a method that returns a typed view into some blob of data for another field) will require us to write our own macro system anyway, I think?

The "mutability for free" thing is interesting, but I'm not sure how well it will work for us in practice. At the very least it probably won't be adequate for manipulating fonts in general, since we can't add or remove things from collections?